File: BOJ/1535.cpp
#include <bits/stdc++.h>
using namespace std;

int n, answer;
int L[20];
int J[20];
int check[20];

void bruteForce(int init, int bital, int happiness) {
	// 기저 조건. 체력이 0이하가 되면 더 이상 진행하지 않고 종료
	// 그렇지 않으면 정답 값을 계속 업데이트 해준다.
	if (bital <= 0) return;
	else answer = max(answer, happiness);

	// 최대 사람 수가 20명이기 때문에 모든 경우를 탐색하는 bruteforce를 사용할 것.
	for (int i = init; i < n; i++) {
		if (!check[i]) { // 방문하지 않은 노드가 있으면 방문 해본다.
			check[i] = 1;
			bruteForce(i + 1, bital - L[i], happiness + J[i]); // 인사를 할 때, 잃는 체력과 얻는 기쁨을 가지고 다음 task로 진행
			check[i] = 0;
		}
	}
}

int main() {
	cin >> n;
	for (int i = 0; i < n; i++) {
		cin >> L[i];
	}
	for (int i = 0; i < n; i++) {
		cin >> J[i];
	}

	bruteForce(0, 100, 0);

	cout << answer << '\n';

	return 0;
}

File: BOJ/python/1535.py
"""

최대 사람 수가 20명이기 때문에 bruteforce를 사용하여 풀 수 있겠다고 생각했다.
기쁨을 얻을 수 있는 모든 경우의 수를 탐색하여 가장 기쁨을 많이 얻은 경우를 출력해주었다.

"""
import sys

input = sys.stdin.readline # line마다 입력을 받는다.

global answer # answer를 전역 변수로 사용한다.
answer = 0
num = int(input())
health = list(map(int, input().split(" ")))
happiness = list(map(int, input().split(" ")))
check = [0 for i in range(20)] # 방문을 check하기 위한 list


def bruteForce(init, bital, happy):
    # 현재 체력이 0이하가 되면 함수를 종료한다. 그렇지 않으면 
    # 현재까지 얻은 기쁨 값중 가장 큰 값을 업데이트 해준다.
    if bital <= 0:
        return
    else:
        global answer
        answer = max(answer, happy)

    for i in range(init, num):
        # 인사하지 않았다면 인사를 해본다.
        if check[i] == 0:
            check[i] = 1
            # 인사하고 얻은 기쁨과 잃는 체력을 가지고 다음 task를 진행한다.
            bruteForce(i + 1, bital - health[i], happy + happiness[i])
            check[i] = 0


bruteForce(0, 100, 0)
print(answer)
